src/reformat.c

```diff
@@ -435,6 +445,17 @@ avifResult avifImageRGBToYUV(avifImage * image, const avifRGBImage * rgb)
     return AVIF_RESULT_OK;
 }
 
+static void avifStoreRGB8Pixel(avifRGBFormat format, uint8_t R, uint8_t G, uint8_t B, uint8_t * ptrR, uint8_t * ptrG, uint8_t * ptrB)
+{
+    if (format == AVIF_RGB_FORMAT_RGB_565) {
+        *(uint16_t *)ptrR = (uint16_t)((B >> 3) | ((G >> 2) << 5) | ((R >> 3) << 11));
```

Comment:
No rounding for faster computation?

Reply:
Sorry i am not sure what you mean by no rounding here. Can you please explain?

Reply:
I think he means dithering here:
https://chromium.googlesource.com/webm/libwebp/+/34b1dc336d04a1b3bb54f38dce2a74e2a586427f/extras/extras.c#50

Reply:
There is dithering yes, but also sometimes there is what I (mistakenly?) call rounding:

`(value >> (1 << shift >> 1)) >> shift` instead of `value >> shift` (flooring) to have lower loss when scaling back to `shifted_value << shift`.
Example: `value=15`, `shift=3` gives `15>>3=1` (flooring), `(15+(1<<2))>>3=2` (rounding), `1<<3=8` (flooring diff `15-8=7`), `2<<3=16` (rounding diff `15-16=-1`).
But maybe the rounding is done elsewhere (such as during the `<<shift`), or maybe it is faster not to take rounding into account. I was just wondering what was the reason.

Reply:
just to be clear, what's happening here is that we are converting R to 5 bits (taking the 5 MSB), G to 6 bits and B to 5 bits. We are then concatenating those bits to be a 16bit word.

i don't think rounding and scaling back to the original value is of much relevance here since RGB565 itself is quite a lossy.

i was trying to match libyuv's C implementation so that we have consistent results between libyuv and libavif.

https://chromium.googlesource.com/libyuv/libyuv/+/9892d70c965678381d2a70a1c9002d1cf136ee78/source/row_common.cc#2362

Reply:
> I think he means dithering here: https://chromium.googlesource.com/webm/libwebp/+/34b1dc336d04a1b3bb54f38dce2a74e2a586427f/extras/extras.c#50

Isn't this for going from RGB565 to something else?

We are going the other way here (i.e.) YUV -> RGB565. So i am not sure how dithering can be applied here.

Reply:
Vignesh is right. (I saw the same dithering code in libyuv's `RGB565ToARGBRow_C` function.)

The rounding that Yannis suggested needs to be done carefully to avoid converting, e.g., a red value of 0xFF to 0x20 (which is a 6-bit value) instead of 0x1F (the maximum 5-bit value).

Reply:
The example I gave was 565->888, but you can dither (add noise to) u & v before converting to rgb:
https://chromium.googlesource.com/webm/libwebp/+/refs/tags/v1.2.4/src/dec/frame_dec.c#371

Reply:
Yeah, so in order to keep things simple i am just going to match libyuv's C conversion.

Marking this conversation as resolved. Please feel free to re-open if anybody things otherwise. :)